app/src/androidTest/java/chat/rocket/android/matchers/CustomMatchers.kt

```diff
@@ -74,8 +77,59 @@ fun withTextInChild(id: Int, text: String): ViewAction {
         }
 
         override fun perform(uiController: UiController, view: View) {
-            val v = view?.findViewById<TextView>(id)
+            val v = view.findViewById<TextView>(id)
             v?.text = text
         }
     }
+}
+
+//Matches an element with an id at a particular index
+fun withIndex(matcher: Matcher<View>, index: Int): Matcher<View> {
+    return object : TypeSafeMatcher<View>() {
+        var currentIndex = 0
+
+        override fun describeTo(description: Description) {
+            description.appendText("with index: ")
+            description.appendValue(index)
+            matcher.describeTo(description)
+        }
+
+        public override fun matchesSafely(view: View): Boolean {
+            return matcher.matches(view) && currentIndex++ == index
+        }
+    }
+}
+
+class ScrollToBottom : ViewAction {
+    override fun getDescription(): String {
+        return "scroll RecyclerView to bottom"
+    }
+
+    override fun getConstraints(): Matcher<View> {
+        return allOf<View>(isAssignableFrom(RecyclerView::class.java), isDisplayed())
+    }
+
+    override fun perform(uiController: UiController?, view: View?) {
+        val recyclerView = view as RecyclerView
+        val itemCount = recyclerView.adapter?.itemCount
+        val position = itemCount?.minus(1) ?: 0
+        recyclerView.scrollToPosition(position)
+        uiController?.loopMainThreadUntilIdle()
+    }
+}
+
+class ScrollToTop : ViewAction {
+    override fun getDescription(): String {
+        return "scroll RecyclerView to bottom"
+    }
+
+    override fun getConstraints(): Matcher<View> {
+        return allOf<View>(isAssignableFrom(RecyclerView::class.java), isDisplayed())
+    }
+
+    override fun perform(uiController: UiController?, view: View?) {
+        val recyclerView = view as RecyclerView
+        recyclerView.scrollToPosition(0)
+        uiController?.loopMainThreadUntilIdle()
+    }
```

Comment:
Maybe this file is your library file where you can put that loginIfUserIsLoggedOut function?

Also, is CustomMatchers the best name for this file?  This file seems to have multiple uses?

Reply:
Yes, It is where all util functions are present. I will move loginIfUserIsLoggedOut function to here and will give a suitable name to the file.